File: services/api/app/schemas/reader_scene.py
from __future__ import annotations

from datetime import datetime
from typing import Any, Literal
from uuid import UUID

from pydantic import BaseModel, Field

from app.schemas.analysis import AnalyzeRequestMeta, ContentSummary, SourceType


class ReaderRecordMeta(BaseModel):
    id: UUID
    client_record_id: str | None = None
    title: str | None = None
    source_type: SourceType
    source_text: str
    request_payload_json: dict[str, Any] = Field(default_factory=dict)
    reading_goal: str | None = None
    reading_variant: str | None = None
    analysis_status: str
    user_facing_state: str | None = None
    workflow_version: str | None = None
    schema_version: str | None = None
    created_at: datetime
    updated_at: datetime


class ReaderArticleParagraph(BaseModel):
    paragraph_id: str
    sentence_ids: list[str] = Field(default_factory=list)


class ReaderArticleSentence(BaseModel):
    sentence_id: str
    paragraph_id: str
    text: str


class ReaderArticle(BaseModel):
    paragraphs: list[ReaderArticleParagraph] = Field(default_factory=list)
    sentences: list[ReaderArticleSentence] = Field(default_factory=list)
    source_text: str | None = None


class ReaderSceneModel(BaseModel):
    schema_version: str
    request: AnalyzeRequestMeta
    article: ReaderArticle
    user_facing_state: str = "normal"
    translations: list[dict[str, Any]] = Field(default_factory=list)
    inline_marks: list[dict[str, Any]] = Field(default_factory=list)
    sentence_entries: list[dict[str, Any]] = Field(default_factory=list)
    warnings: list[dict[str, Any]] = Field(default_factory=list)
    content_summary: ContentSummary | None = None
    title: str | None = None


class ReaderViewMeta(BaseModel):
    view_version: str
    data_source: Literal["render_scene_snapshot", "source_text_fallback"]
    fallback_mode: Literal["none", "article_rebuilt_from_source_text", "scene_missing"]
    supplements_merged: bool


class ReaderSceneResponse(BaseModel):
    record_meta: ReaderRecordMeta
    reader_scene: ReaderSceneModel
    view_meta: ReaderViewMeta

File: apps/web/src/types/api/reader-scene.ts
export interface ReaderSceneRecordMetaDto {
  id: string;
  client_record_id: string | null;
  title: string | null;
  source_type: "user_input" | "daily_article" | "imported" | "ocr";
  source_text: string;
  request_payload_json: Record<string, unknown>;
  reading_goal: string | null;
  reading_variant: string | null;
  analysis_status: string;
  user_facing_state: string | null;
  workflow_version: string | null;
  schema_version: string | null;
  created_at: string;
  updated_at: string;
}

export interface ReaderSceneDto {
  schema_version: string;
  request: Record<string, unknown>;
  article: Record<string, unknown>;
  user_facing_state: string;
  translations: Record<string, unknown>[];
  inline_marks: Record<string, unknown>[];
  sentence_entries: Record<string, unknown>[];
  warnings: Record<string, unknown>[];
  content_summary?: Record<string, unknown> | null;
  title?: string | null;
}

export interface ReaderSceneViewMetaDto {
  view_version: string;
  data_source: "render_scene_snapshot" | "source_text_fallback";
  fallback_mode: "none" | "article_rebuilt_from_source_text" | "scene_missing";
  supplements_merged: boolean;
}

export interface ReaderSceneResponseDto {
  record_meta: ReaderSceneRecordMetaDto;
  reader_scene: ReaderSceneDto;
  view_meta: ReaderSceneViewMetaDto;
}
